pkg/route/secret/monitor_test.go

```diff
@@ -0,0 +1,354 @@
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Data: map[string][]byte{
+			"test": {1, 2, 3, 4},
+		},
+	}
+}
+
+func TestStartInformer(t *testing.T) {
+	scenarios := []struct {
+		name      string
+		isClosed  bool
+		expectErr bool
+	}{
+		{
+			name:      "pass closed channel into informer",
+			isClosed:  true,
+			expectErr: true,
+		},
+		{
+			name:      "pass unclosed channel into informer",
+			isClosed:  false,
+			expectErr: false,
+		},
+	}
+
+	for _, s := range scenarios {
+		t.Run(s.name, func(t *testing.T) {
+			fakeKubeClient := fake.NewSimpleClientset()
+			monitor := fakeMonitor(context.TODO(), fakeKubeClient, ObjectKey{})
+			if s.isClosed {
+				close(monitor.stopCh)
+			}
+			go monitor.StartInformer(context.TODO())
+			time.Sleep(10 * time.Millisecond)
+
+			select {
+			// this case will execute if stopCh is closed
+			case <-monitor.stopCh:
+				if !s.expectErr {
+					t.Error("informer is not running")
+				}
+			default:
+				t.Log("informer is running")
+			}
+		})
+	}
+}
+
+func TestStopInformer(t *testing.T) {
+	scenarios := []struct {
+		name             string
+		withStart        bool
+		alreadyStopped   bool
+		expectStopped    bool
+		expectChanClosed bool
+	}{
+		{
+			name:             "stop without starting informer",
+			withStart:        false,
+			expectStopped:    false,
+			expectChanClosed: false,
+		},
+		{
+			name:             "stopping already stopped informer",
+			withStart:        true,
+			alreadyStopped:   true,
+			expectStopped:    false,
+			expectChanClosed: true,
+		},
+		{
+			name:             "correctly stopped informer",
+			withStart:        true,
+			alreadyStopped:   false,
+			expectStopped:    true,
+			expectChanClosed: true,
+		},
+	}
+
+	for _, s := range scenarios {
+		t.Run(s.name, func(t *testing.T) {
+			fakeKubeClient := fake.NewSimpleClientset()
+			monitor := fakeMonitor(context.TODO(), fakeKubeClient, ObjectKey{})
+
+			if s.withStart {
+				go monitor.StartInformer(context.TODO())
+				time.Sleep(10 * time.Millisecond)
+			}
+			if s.alreadyStopped {
+				monitor.StopInformer()
+			}
+
+			if monitor.StopInformer() != s.expectStopped {
+				t.Error("unexpected result")
+			}
+
+			var chanClosed bool
+			select {
+			// this case will execute if stopCh is closed
+			case <-monitor.stopCh:
+				chanClosed = true
+			default:
+				chanClosed = false
+			}
+
+			if s.expectChanClosed != chanClosed {
+				t.Error("unexpected result on chan closure")
+			}
+		})
+	}
+}
+
+func TestStopWithContextCancel(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.TODO())
+	fakeKubeClient := fake.NewSimpleClientset()
+	monitor := fakeMonitor(ctx, fakeKubeClient, ObjectKey{})
+
+	go monitor.StartInformer(ctx)
+	time.Sleep(10 * time.Millisecond)
+
+	// cancel the context to stop the informer
+	cancel()
+	time.Sleep(10 * time.Millisecond)
```

Comment:
Can  we avoid doing sleeps by trying  something like this?
```suggestion
	go func () {
	      started <- true
	      monitor.StartInformer(ctx)
	 }()
	
	<-started
	// cancel the context to stop the informer
	cancel()
```
Or do we ever need all this? Can't we just run `StartInformer` as a goroutine and call `cancel()` in the main thread?

Reply:
`StartInformer` is a blocking call, so I'm now utilizing `HasSynced` to signal out that the informer has started running. This is what it looks like now, let me know if it's better

```go
        stopCh := make(chan bool)

	go func() {
		monitor.StartInformer(ctx)
		close(stopCh)
	}()
	if !cache.WaitForCacheSync(ctx.Done(), monitor.HasSynced) {
		t.Fatal("cache not synced yet")
	}

	// cancel the context to stop the informer
	cancel()
	<-stopCh
```

Reply:
Yes, that's definitely more deterministic than sleeps. Thanks!